README.rst
=======================================
Conversate - Persistent chat for Django
=======================================

A simple lightweight persistent chat app for Django sites, designed for users who may
not always be around at the same time.

* Project site: http://radiac.net/projects/django-conversate/
* Source code: https://github.com/radiac/django-conversate

.. image:: https://travis-ci.org/radiac/django-conversate.svg?branch=master
    :target: https://travis-ci.org/radiac/django-conversate

.. image:: https://coveralls.io/repos/radiac/django-conversate/badge.svg?branch=master&service=github
    :target: https://coveralls.io/github/radiac/django-conversate?branch=master


Features
========

* Admins create rooms and add users to them
* Messages are stored in the database
* Full history is available to all users of that room
* File upload support
* Simple jquery-based polling for real-time updates
* Users can opt in to receive e-mail alerts of activity when away
* Support for users without javascript

Supports Django 2.2+, on Python 3.6+.

* See `CHANGES <CHANGES>`_ for full changelog and roadmap
* See `UPGRADE <UPGRADE.rst>`_ for how to upgrade from earlier releases

Please note: this is designed for small groups to have infrequent conversations
which are often asynchronous; because each poll event will create a new HTTP
connection, installations where a large number of concurrent users are expected
should look for a different solution involving long-polling and websockets.

There is an example site in the ``example`` directory.


Installation
============

1. Install ``django-conversate``::

    pip install django-conversate

2. Add Conversate to ``INSTALLED_APPS``::

    INSTALLED_APPS = (
        ...
        'conversate',
    )

   You may also want to change some settings here (see `Settings`_ below)

3. Include the URLconf in your project's urls.py::

    url(r'^conversate/', include('conversate.urls', namespace='conversate')),

4. Make sure your ``base.html`` template has the necessary blocks, or override
   Conversate's base, ``conversate/base.html`` (see `Templates and styles`_ below). You
   will also want to create a link somewhere to ``conversate-index`` (or
   ``conversate.views.index``) so users can access it.

5. Add the models to the database::

    python manage.py migrate conversate


Settings
--------

Add these settings to your ``settings.py`` file to override the defaults.

``CONVERSATE_PAGE_SIZE``
    Number of lines to show on a page

    Default: ``100``

``CONVERSATE_IDLE_AT``:
    The time until the UI decides that the user is idle (in ms)

    Default: ``60*1000``

``CONVERSATE_POLL_MIN``:
    Minimum poll interval (in ms)

    Default: ``5 * 1000``

``CONVERSATE_POLL_MAX``:
    Maximum poll interval (in ms)

    Default: ``60 * 1000``

``CONVERSATE_POLL_STEP``:
    Amount to increase poll interval by when there is no activity (in ms)

    Default: ``5 * 1000``

``CONVERSATE_CONTROL_LAYOUT``:
    If True, Conversate's JavaScript will control the layout:
    * changes the container element to position:relative and padding:0
    * maximises container element's height into available space in the window
    * makes the conversation scroll in place
    * moves the input field to the bottom of the container

    Default: ``True``

``CONVERSATE_DISCONNECT_AT``:
    How long before marking the user as disconnected (in secs)

    Defaults to POLL_MAX plus 30 seconds, ``60 + 30``

``CONVERSATE_EMAIL_FROM``:
    From address for alert e-mails

    Default: ``DEFAULT_FROM_EMAIL`` (from main Django settings)

``CONVERSATE_STORE_ROOT``:
    Path to dir where uploaded files are stored.

    This should not be publicly accessible otherwise files could be downloaded without
    permission. The default value is the media root so that it works out of the box, but
    this is insecure as permission checks can be bypassed.

    Example: ``os.path.join(BASE_DIR, "..", "private")``

    Default: ``MEDIA_ROOT``


Templates and styles
--------------------

The Conversate templates extend ``conversate/base.html``, which in turn extends
``base.html``. The templates use HTML5 elements.

They will expect the following blocks:

* ``js`` for inserting JavaScript
* ``css`` for inserting CSS
* ``title`` for inserting the title (plain text) - or ``{{ title }}`` instead
* ``content`` for the body content

You will need to add these to your ``base.html`` template. Alternatively, if
you already have the blocks but with different names, create
``conversate/base.html`` in your own templates folder and map them; for
example::

    {% block script %}
        {{ block.super }}
        {% block js %}{% endblock %}
    {% endblock %}

Once you have mapped these blocks, the default settings and templates should
work out of the box with most designs. However, the conversate container
element in your site's base template should be given a fixed height and width
to contain the chat interface.

There is a single global JavaScript variable used, ``CONVERSATE``, which the
template uses to pass settings and variables to the JavaScript.


Usage
=====

Set up one or more rooms in the Django admin site, and the rooms will be listed
for your users on the conversate index page.

Users can double-click the poll timer to force a faster poll.


Credits
=======

Thanks to all contributors, who are listed in CHANGES.

This project includes bundled JavaScript dependencies.


CHANGES
=========================
Django Conversate Changes
=========================


Upgrading
=========

Upgrading from 0.1.0
--------------------

* Add a namespace to your url import.
* Only tested on Django 1.11 under Python 3.5. Patches to improve compatibility
  are welcome.
* Styles have changed, you may need to update your templates


Changelog
=========

0.3.1       2020-09-13
            Change:     Documentation

0.3.0       2020-09-13
            Feature:    Supports file uploads
            Change:     Supports Django 2.2+
            Change:     Requires Python 3.6
            Bugfix:     Server-side API errors return gracefully
            Security:   Upstream JS fixes

0.2.0       2018-11-18
            Feature:    Supports multi-line entry (shift+enter) and markdown
            Change:     Minimum Django requirement now 1.11
            Change:     Requires Python 3.5
            Change:     Urls now namespaced
            Change:     Times are now timezone aware
            Change:     South migrations are removed
            Change:     Layout simplified to use flexbox

0.1.0       2015-08-07
            Feature:    Initial release


Roadmap
=======

* Support for websockets
* Documented API for integration with third party apps
* Improve room.updateTimes to update older entries
* UI improvements
  * Room selection sidebar
  * API to automatically save settings changes
* Admin options
  * Let users edit and delete their messages
  * Clear room without deleting it
  * UI for creating and managing rooms
  * Admin option to make a room public, so users can add themselves
* User options
    * Separate flash and title visual alerts into two options
    * Provide custom e-mail address for alerts
    * Set alert time ranges
    * Colour picker

The roadmap is subject to change; let me know if something is important to you.


conversate/settings.py
from django.conf import settings


# Number of lines to show on a page
PAGE_SIZE = getattr(settings, "CONVERSATE_PAGE_SIZE", 100)

# Time until UI decides user is idle (in ms)
IDLE_AT = getattr(settings, "CONVERSATE_IDLE_AT", 60 * 1000)

# Minimum poll interval (in ms)
POLL_MIN = getattr(settings, "CONVERSATE_POLL_MIN", 5 * 1000)

# Maximum poll interval (in ms)
POLL_MAX = getattr(settings, "CONVERSATE_POLL_MAX", 60 * 1000)

# Amount to increase poll interval by when there is no activity (in ms)
POLL_STEP = getattr(settings, "CONVERSATE_POLL_STEP", 5 * 1000)

# If True, Conversate's JavaScript will control the layout:
#   - changes the container element to position:relative and padding:0
#   - maximises container element's height into available space in the window
#   - makes the conversation scroll in place
#   - moves the input field to the bottom of the container
CONTROL_LAYOUT = getattr(settings, "CONVERSATE_CONTROL_LAYOUT", True)

# How long before marking the user as disconnected (in secs)
# Defaults to POLL_MAX plus 30 seconds
DISCONNECT_AT = getattr(
    settings,
    "CONVERSATE_DISCONNECT_AT",
    (POLL_MAX / 1000) + 30,
)

# From address
EMAIL_FROM = getattr(
    settings,
    "CONVERSATE_EMAIL_FROM",
    getattr(settings, "DEFAULT_FROM_EMAIL"),
)

# Path to the private file store
#
# This should be a private dir outside the main site structure.
#
# Defaults to media root so it will work out of the box, but this has security
# implications - permission checks can be bypassed.
# Default: MEDIA_ROOT
STORE_ROOT = getattr(settings, "CONVERSATE_STORE_ROOT", settings.MEDIA_ROOT)
